R/PipeOpVtreat.R

```diff
@@ -0,0 +1,217 @@
+#'
+#' pop = PipeOpVtreat$new()
+#' pop$train(list(task))
+#' @family PipeOps
+#' @include PipeOpTaskPreproc.R
+#' @export
+PipeOpVtreat = R6Class("PipeOpVtreat",
+  inherit = PipeOpTaskPreproc,
+  public = list(
+    initialize = function(id = "vtreat", param_vals = list()) {
+      ps = ParamSet$new(params = list(
+        ParamLgl$new("recommended", default = TRUE, tags = c("train", "predict")),
+        ParamUty$new("cols_to_copy", default = NULL, tags = "train", custom_check = function(x) checkmate::check_list(x, null.ok = TRUE)),
+        ParamUty$new("params", default = NULL, tags = "train", custom_check = function(x) checkmate::check_list(x, null.ok = TRUE)),
+        ParamUty$new("imputation_map", default = NULL, tags = "train", custom_check = function(x) checkmate::check_list(x, null.ok = TRUE))
+        # FIXME: parallelCluster
+      ))
+      ps$values = list(recommended = TRUE, cols_to_copy = NULL, params = NULL, imputation_map = NULL)
+      super$initialize(id, param_set = ps, param_vals = param_vals, packages = "vtreat")
+    }
+  ),
+  private = list(
+
+    .train_task = function(task) {
+
+      var_list = task$feature_names
+      if (length(var_list) == 0) {
+        return(task)  # early exit
+      }
+
+      task_type = task$task_type
+
+      if (task_type == "regr") {
+        treatment = vtreat::NumericOutcomeTreatment
+        params = vtreat::regression_parameters(self$param_set$values$params)
+      } else if (task_type == "classif") {
+        if (length(task$class_names) > 2L) {
+          treatment = vtreat::MultinomialOutcomeTreatment
+          params = vtreat::multinomial_parameters(self$param_set$values$params)
+        } else {
+          treatment = vtreat::BinomialOutcomeTreatment
+          params = vtreat::classification_parameters(self$param_set$values$params)
+        }
+      }
+
+      if (length(self$param_set$values$cols_to_copy)) {
+        checkmate::assert_subset(unlist(self$param_set$values$cols_to_copy), choices = var_list, empty.ok = TRUE)
+      }
+
+      if (length(self$param_set$values$imputation_map)) {
+        checkmate::assert_subset(names(self$param_set$values$imputation_map), choices = var_list, empty.ok = TRUE)
+      }
+
+      transform_design = mlr3misc::invoke(treatment,
+        var_list = var_list,
+        outcome_name = task$target_names,
+        cols_to_copy = self$param_set$values$cols_to_copy,
+        params = params,
+        imputation_map = self$param_set$values$imputation_map)
+
+      # the following exception handling is necessary because vtreat sometimes fails with "no usable vars" if the data is already "clean" enough
+      vtreat_res = try(
+        mlr3misc::invoke(vtreat::fit_prepare,
+          vps = transform_design,
+          dframe = task$data(),
+          weights = task$weights$weight),
+        silent = TRUE
+      )
+
+      if (class(vtreat_res) == "try-error") {
+        self$state = NULL
```

Comment:
self$state must not be `NULL`, use `list(NULL)` or `list()` or something like that.